datatrails: fix bookmark/recent trail detection, prevent duplications, save trail on browser close/reload

darrenjaneczek opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

- fixes detection on bookmarks and recents when current step isn't the last step on the trail history
- now saves current trail on browser close or reload (unload)
- refresh page or return to recent trail URL will restore state to that trail and allow user to resume
  that trail instead of creating a duplicate recent trail
- do not create a recent trail out of an empty starting trail

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

darrenjaneczek committed:
fix: persistence trail detection, save on unload

- fixes detection on bookmarks and recents when current step isn't final
- now save current trail on browser close or reload (unload)
- refresh page or return to URL that corresponds to a recent trail will
   resume that trail instead of creating a duplicate recent trail
- do not create a recent trail out of an empty starting trail

torkelo reviewed (approved):
Nice tests! 

The bookmark button state should cleared after any change is made as the bookmark is only for that specific state. And if possible clear the history when we create the bookmark.

darrenjaneczek committed:
fix: bookmarks status after making new step

- clone bookmark trail state to prevent it from being changed by user
- re-evaluate bookmark status after creating new step

darrenjaneczek commented:
@torkelo 
> Nice tests!
> 
> The bookmark button state should cleared after any change is made as the bookmark is only for that specific state. And if possible clear the history when we create the bookmark.

Resolved the bookmark button state being kept after changing state via:
- https://github.com/grafana/grafana/pull/85677/commits/01ba40b2da31d30c728e1aac6676a47e7da642d1

Merging in the fix from:
- https://github.com/grafana/grafana/pull/86843 
also helps with bookmark stability by not retroactively changing the parameter state of steps.

I had a local stash saved to only store the bookmark state vs. the whole history. It is not in the following draft PR is here for consideration:
- https://github.com/grafana/grafana/pull/86929

darrenjaneczek committed:
fix: comments

darrenjaneczek committed:
fix: use valid filters value in unit test

torkelo commented on public/app/features/trails/DataTrailsApp.tsx:
```diff
@@ -99,8 +91,23 @@ let dataTrailsApp: DataTrailsApp;
   if (!dataTrailsApp) {
+    const newTrail = newMetricsTrail();
+
+    // Set the initial state of the newTrail based on the URL,
+    // In case we are initializing from an externally created URL or a page reload
+    getUrlSyncManager().initSync(newTrail);
+    // Remove the URL sync for now. It will be restored on the trail if it is activated.
+    getUrlSyncManager().cleanUp(newTrail);
```
is this really needed?

darrenjaneczek replied:
Yes, in order to find the "recentMatchingTrail" below.

This is part of resolving the "continue same trail on refresh" bug.

darrenjaneczek replied:
(unless there is a more straightforward way of populating the trail state from the current url)

torkelo replied:
@darrenjaneczek continue same trail on refresh? I refresh should clear the history

darrenjaneczek replied:
It was deemed to be a bug.
With this change, accessing the same URL, either by refresh, or explicit URL, will act similar to selecting the trail from the recent list.

torkelo reviewed (approved):
Looks ok.

Still confused why we store the full state in bookmark vs just the URL but a separate topic :)

darrenjaneczek committed:
fix: clarified the initial trail
